### PR TITLE
Ranks

### PR DESCRIPTION
**Pull request by maikelbos0**
Part one, just added ranks. Not really sure about having a separate little service class for calculating rank but I felt like keep adding to Player class is a bad idea too. Any ideas?

Also rank gets updated with the turn, not in between. IDK about that but it's more efficient.

I'm not worried about any imbalance for the actual rank right now, but the structure is important to get right.

### CODE REVIEW

File: WarOfEmpires.Domain/Players/Player.cs [hunk 1]
```diff
@@ -36,6 +36,7 @@ public class Player : AggregateRoot {
         public virtual int BankTurns { get; protected set; } = 6;
         public virtual int Stamina { get; protected set; } = 100;
         public virtual bool HasUpkeepRunOut { get; protected set; } = false;
+        public virtual int Rank { get; protected set; } = int.MaxValue;
```
Comment (KatarinaSundqvist) [inline WarOfEmpires.Domain/Players/Player.cs:39]: Question: Does this mean that new players would initially be max ranked, until their rank is processed the first time?

**Comment by KatarinaSundqvist** [comment]
Updating only with the turn is risky, and encourages people to make sure they're doing ok at the turn, or attacking your rival (e.g. in a GO contest) just before the turn, so they don't have time to recover. But continuous updating has some resource drawbacks, I guess.

Agreed that the Player class has grown large. Could a sensible approach be to break out some sub classes, e,g, PlayerTroops, PlayerWorkers, PlayerBuildings or similar, where the respective methods could live?

File: WarOfEmpires.Domain/Players/Player.cs [hunk 2]
```diff
@@ -184,9 +185,10 @@ public class Player : AggregateRoot {
...
         }
 
-        public virtual void ProcessTurn() {
+        public virtual void ProcessTurn(int newRank) {
             var upkeep = GetUpkeepPerTurn();
 
+            Rank = newRank;
             AttackTurns++;
 
             if (Stamina < 100) {
```
Comment (KatarinaSundqvist) [inline WarOfEmpires.Domain/Players/Player.cs:194]: I can't comment on line 299, but please correct it. It currently says "// Adjust for havnig not enough soldiers"
Please change it to "// Adjust for not having enough soldiers"

File: WarOfEmpires.Domain/Players/Player.cs (hunk 1 above)
Comment (maikelbos0) [reply WarOfEmpires.Domain/Players/Player.cs:39]: Bottom ranked; 1 is the max

File: WarOfEmpires.Domain/Players/Player.cs (hunk 2 above)
Comment (maikelbos0) [reply WarOfEmpires.Domain/Players/Player.cs:194]: Done

**Comment by maikelbos0** [comment]
> Updating only with the turn is risky, and encourages people to make sure they're doing ok at the turn, or attacking your rival (e.g. in a GO contest) just before the turn, so they don't have time to recover. But continuous updating has some resource drawbacks, I guess.
> 
> Agreed that the Player class has grown large. Could a sensible approach be to break out some sub classes, e,g, PlayerTroops, PlayerWorkers, PlayerBuildings or similar, where the respective methods could live?

**Comment by maikelbos0** [comment]
I've moved the rank update to its own scheduled task that's triggered (for now) every 2 minutes. I've also created a separate event for it, I think that also gives us the possibility to invoke it directly when for instance the GO gets attacked - but that's a later worry.

**Pull request merged**